File: scala/src/main/scala/euler/Sieve.scala
/* Prime number sieve. */

package euler

import scala.collection.mutable.BitSet
import scala.collection.mutable.ListBuffer

// An integer based sieve.  Simple imperative approach.
class Sieve(val limit: Int) {
  val composite = new BitSet(limit)
  composite += 0
  composite += 1

  def init() {
    var p = 2
    while (p < limit) {
      var n = p+p
      while (n < limit) {
        composite += n
        n += p
      }

      p = if (p == 2) 3 else p+2
    }
  }
  init()

  def isPrime(num: Int): Boolean = !composite.contains(num)
}

case class Factor(prime: Int, power: Int)

class AutoSieve {

  def isPrime(num: Int): Boolean = {
    if (num >= sieve.limit) {
      sieve = new Sieve(findLimit(num))
    }
    sieve.isPrime(num)
  }

  def nextPrime(num: Int): Int = {
    var p = if (num == 2) 3 else num+2
    while (!isPrime(p))
      p += 2
    p
  }

  def factorize(number: Int): List[Factor] = {
    var num = number
    val result = new ListBuffer[Factor]

    var p = 2
    var count = 0
    while (num > 1) {
      if (num % p == 0) {
        num /= p
        count += 1
      } else {
        if (count > 0) {
          result += Factor(p, count)
          count = 0
        }
        p = nextPrime(p)
      }
    }
    if (count > 0) {
      result += Factor(p, count)
    }

    result.result
  }

  def divisors(number: Int) = spread(factorize(number)).sorted

  private def spread(factors: List[Factor]): List[Int] = factors match {
    case Nil => List(1)
    case x::others => {
      val rest = spread(others)

      var power = 1
      val result = new ListBuffer[Int]
      for (i <- 0 to x.power) {
        for (elt <- rest)
          result += elt * power

        if (i < power)
          power *= x.prime
      }
      result.result
    }
  }

  def divisorCount(number: Int): Int = {
    val facts = factorize(number)
    facts.foldLeft(1)((a, b) => a * (b.power + 1))
  }

  private var sieve = new Sieve(1024)
  private def findLimit(num: Int): Int = {
    var size = 1024
    while (size < num)
      size *= 8
    size
  }
}

File: go/src/euler/sieve.go
// A basic self-expanding prime sieve.

package euler

import "sort"

type Sieve struct {
	composites []bool
}

func (s *Sieve) IsPrime(n int) bool {
	size := 8192

	if s.composites != nil && len(s.composites) <= n {
		size = len(s.composites)
		s.composites = nil
	}

	if s.composites == nil {
		for size <= n {
			size *= 8
		}
		s.build(size)
	}

	return !s.composites[n]
}

func (s *Sieve) build(size int) {
	comp := make([]bool, size)
	comp[0] = true
	comp[1] = true

	pos := 2
	for pos < size {
		if comp[pos] {
			pos += 2
		} else {
			n := pos + pos
			for n < size {
				comp[n] = true
				n += pos
			}
			if pos == 2 {
				pos += 1
			} else {
				pos += 2
			}
		}
	}
	s.composites = comp
}

func (s *Sieve) NextPrime(base int) (next int) {
	if base == 2 {
		next = 3
	} else {
		next = base + 2
	}

	for !s.IsPrime(next) {
		next += 2
	}

	return
}

type Factor struct {
	Prime int
	Power int
}

func (s *Sieve) Factorize(num int) (factors []Factor) {
	factors = make([]Factor, 0, 5)

	count := 0
	prime := 2
	for num > 1 {
		if (num % prime) == 0 {
			num /= prime
			count++
		} else {
			factors = append(factors, Factor{Prime: prime, Power: count})
			count = 0

			if num > 1 {
				prime = s.NextPrime(prime)
			}
		}
	}
	if count > 0 {
		factors = append(factors, Factor{Prime: prime, Power: count})
	}

	return
}

func (s *Sieve) Divisors(num int) (result []int) {
	result = s.spread(s.Factorize(num))
	sort.Ints(result)
	return
}

func (s *Sieve) spread(factors []Factor) (result []int) {
	if len(factors) == 0 {
		result = []int{1}
		return
	}
	x := factors[0]
	rest := s.spread(factors[1:])
	result = make([]int, 0, 8)
	power := 1
	for i := 0; i <= x.Power; i++ {
		for _, rs := range rest {
			result = append(result, rs*power)
		}
		power *= x.Prime
	}
	return
}

func (s *Sieve) ProperDivisorSum(num int) (result int) {
	divs := s.Divisors(num)
	for _, d := range divs {
		result += d
	}
	result -= num
	return
}
